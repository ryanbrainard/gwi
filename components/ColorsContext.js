import React from "react";

export default (ColorsContext = React.createContext());

export function provideColors(colors, Component) {
  return props => {
    return (
      <ColorsContext.Provider value={colors}>
        <Component {...props} />
      </ColorsContext.Provider>
    );
  };
}

export function withColors(Component) {
  return props => {
    return (
      <ColorsContext.Consumer>
        {colors => <Component {...props} colors={colors} />}
      </ColorsContext.Consumer>
    );
  };
}
