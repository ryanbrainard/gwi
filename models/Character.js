import VoicePlayerMRU from "./VoicePlayerMRU";

export default class Character {
  static _chars = {};

  static register(char) {
    this._chars[char.name] = char;
    return char;
  }

  static find(name) {
    return this._chars[name];
  }

  constructor(name, group, voices) {
    this._name = name;
    this._group = group;
    this._voices = voices;

    this.player = new VoicePlayerMRU(this._voices);
  }

  get key() {
    return this.name;
  }

  get name() {
    return this._name;
  }

  get group() {
    return this._group.map(c => Character.find(c));
  }

  async preloadPlay() {
    await this.player.load();
  }

  async play() {
    await this.player.play();
  }
}
