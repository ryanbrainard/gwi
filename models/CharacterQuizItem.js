export default class CharacterQuizItem {
  constructor(char, stateSetter) {
    this._char = char;
    this._stateSetter = stateSetter;
    this._success = false;
  }

  get character() {
    return this._char;
  }

  get success() {
    return this._success;
  }

  set success(value) {
    this._success = value;
    this._stateSetter(this); // TODO: good idea or just gross?
  }
}
