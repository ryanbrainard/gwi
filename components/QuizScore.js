import { Feather, MaterialCommunityIcons } from "@expo/vector-icons";
import PropTypes from "prop-types";
import React from "react";
import { StyleSheet, Text, TouchableOpacity, View } from "react-native";
import config from "../config";
import CharacterQuizItem from "../models/CharacterQuizItem";

export default class QuizScore extends React.PureComponent {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.instanceOf(CharacterQuizItem))
      .isRequired,
    restartQuiz: PropTypes.func.isRequired
  };

  render() {
    const { items, restartQuiz } = this.props;

    const numSuccess = items.reduce((sum, item) => {
      return sum + (item.success ? 1 : 0);
    }, 0);

    const numAnswered = items.reduce((sum, item) => {
      return sum + (item.answered ? 1 : 0);
    }, 0);

    const percSuccess =
      numAnswered === 0 ? 0 : Math.round((numSuccess / numAnswered) * 100);

    const difficultChars = Object.entries(
      items.reduce((stats, item) => {
        if (item.answered && !item.success) {
          stats[item.char.name] = (stats[item.char.name] || 0) + 1;
        }
        return stats;
      }, {})
    )
      .sort((a, b) => b[1] - a[1])
      .map(e => e[0]);

    return (
      <View style={styles.container}>
        <Text style={styles.titleText}>{percSuccess}%</Text>
        <Text>
          ({numSuccess} / {numAnswered})
        </Text>
        <Text>{}</Text>
        <Text>{}</Text>
        {difficultChars.length > 0 && (
          <View style={styles.difficultCharsContainer}>
            <Text>Most difficult characters:</Text>
            <Text>{difficultChars.slice(0, 5).join(", ")}</Text>
            <Text>{}</Text>
            <Text>{}</Text>
          </View>
        )}
        <TouchableOpacity style={styles.button} onPress={restartQuiz}>
          <Text style={styles.buttonText}>Try Again</Text>
        </TouchableOpacity>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center"
  },
  difficultCharsContainer: {
    alignItems: "center"
  },
  titleText: {
    fontSize: 24,
    fontWeight: "bold"
  },
  button: {
    backgroundColor: config.colors.about.primary,
    borderRadius: 75,
    padding: 12,
    justifyContent: "center",
    margin: 10
  },
  buttonText: {
    fontWeight: "bold",
    fontSize: 14,
    color: config.colors.text
  }
});
