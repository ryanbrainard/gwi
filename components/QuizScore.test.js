import React from "react";
import renderer from "react-test-renderer";
import Character from "../models/Character";
import CharacterQuizItem from "../models/CharacterQuizItem";
import { provideColors } from "./ColorsContext";
import QuizScore from "./QuizScore";

test("renders correctly", () => {
  const _QuizScore = provideColors({}, QuizScore);

  const group = ["A", "B", "C"];
  group.forEach(char => Character.register(new Character(char, group)));
  const stateSetter = jest.fn();
  const newItem = (char, answered) => {
    const item = new CharacterQuizItem(
      char,
      stateSetter,
      () => "00000000-0000-0000-0000-000000000000",
      s => s
    );
    item.answered = answered;
    return item;
  };
  const items = [
    newItem(Character.find("A"), undefined), // unanswered
    newItem(Character.find("B"), Character.find("A")), // incorrect
    newItem(Character.find("B"), Character.find("B")), // correct,
    newItem(Character.find("C"), Character.find("A")), // incorrect
    newItem(Character.find("C"), Character.find("A")) // incorrect again
  ];
  items[1].answered = items[0].char;

  const tree = renderer
    .create(<QuizScore items={items} restartQuiz={jest.fn()} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
