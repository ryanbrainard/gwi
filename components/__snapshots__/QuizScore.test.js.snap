// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 24,
        "fontWeight": "bold",
      }
    }
  >
    25
    %
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  >
    (
    1
     / 
    4
    )
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
  />
  <View
    style={
      Object {
        "alignItems": "center",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Most difficult characters:
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      C, B
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    />
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#3CB371",
        "borderRadius": 75,
        "justifyContent": "center",
        "margin": 10,
        "opacity": 1,
        "padding": 12,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#000000",
          "fontSize": 14,
          "fontWeight": "bold",
        }
      }
    >
      Try Again
    </Text>
  </View>
</View>
`;
